File: Core/automation/lib/python/core/items.py
"""
This module allows runtime creation and removal of items. It will also remove
any links from an Item before it is removed. This module requires the
JythonItemProvder and JythonItemChannelLinkProvider component scripts.
"""

from core.jsr223 import scope
scope.scriptExtension.importPreset(None)

import core
from core import osgi
from core.log import logging, LOG_PREFIX
from core.links import remove_all_links

ItemBuilderFactory = osgi.get_service(
        "org.openhab.core.items.ItemBuilderFactory"
    ) or osgi.get_service(
        "org.eclipse.smarthome.core.items.ItemBuilderFactory"
    )

ManagedItemProvider = osgi.get_service(
        "org.openhab.core.items.ManagedItemProvider"
    ) or osgi.get_service(
        "org.eclipse.smarthome.core.items.ManagedItemProvider"
    )

log = logging.getLogger("{}.core.items".format(LOG_PREFIX))

__all__ = ["add_item", "remove_item"]

def add_item(item_or_item_name, item_type=None, category=None, groups=None, label=None, tags=[], gi_base_type=None, group_function=None):
    try:
        if not isinstance(item_or_item_name, basestring) and not hasattr(item_or_item_name, 'name'):
            raise Exception("\"{}\" is not a string or Item".format(item_or_item_name))
        item = item_or_item_name
        if isinstance(item_or_item_name, basestring):
            item_name = item_or_item_name
            if item_type is None:
                raise Exception("Must provide item_type when creating an Item by name")

            base_item = None if item_type != "Group" or gi_base_type is None else ItemBuilderFactory.newItemBuilder(gi_base_type, item_name + "_baseItem").build()
            group_function = None if item_type != "Group" else group_function
            item = ItemBuilderFactory.newItemBuilder(item_type, item_name)\
                                                    .withCategory(category)\
                                                    .withGroups(groups)\
                                                    .withLabel(label)\
                                                    .withBaseItem(base_item)\
                                                    .withGroupFunction(group_function)\
                                                    .withTags(set(tags))\
                                                    .build()

        ManagedItemProvider.add(item)
        log.debug("Item added: [{}]".format(item))
        return item
    except:
        import traceback
        log.error(traceback.format_exc())
        return None

def remove_item(item_or_item_name):
    try:
        item = remove_all_links(item_or_item_name)
        if item is None:
            log.warn("Item cannot be removed because it does not exist in the ItemRegistry: [{}]".format(item_or_item_name))
            return None

        ManagedItemProvider.remove(item.name)
        if scope.itemRegistry.getItems(item.name) == []:
            log.debug("Item removed: [{}]".format(item.name))
            return item
        else:
            log.warn("Failed to remove Item from the ItemRegistry: [{}]".format(item.name))
            return None
    except:
        import traceback
        log.error(traceback.format_exc())
        return None

File: Core/automation/lib/javascript/core/items.js
/**
 * Functions for creating rules
 * 
 * Copyright (c) 2019 Contributors to the openHAB Scripters project
 * 
 * @author Jonathan Gilbert - initial contribution
 */

var OPENHAB_CONF = Java.type("java.lang.System").getenv("OPENHAB_CONF");
load(OPENHAB_CONF+'/automation/lib/javascript/core/osgi.js');


(function(context) {
    'use strict';	

var itemBuilderFactory = get_service(
        "org.openhab.core.items.ItemBuilderFactory"
    ) || get_service(
        "org.eclipse.smarthome.core.items.ItemBuilderFactory"
    )

var managedItemProvider = get_service(
        "org.openhab.core.items.ManagedItemProvider"
    ) || get_service(
        "org.eclipse.smarthome.core.items.ManagedItemProvider"
    )

var HashSet = Java.type("java.util.HashSet");;


    /*
        itemName (str): Item name for the Item to create
        itemType (str): (optional) the type of the Item
        category (str): (optional) the category (icon) for the Item
        groups (str): (optional) a list of groups the Item is a member of
        label (str): (optional) the label for the Item
        tags (list): (optional) a list of tags for the Item
        giBaseType (str): (optional) the group Item base type for the Item
        groupFunction (GroupFunction): (optional) the group function used by the Item
            
        returns: the item object, or undefined otherwise  
    */
context.addItem = function(itemName, itemType, category, groups, label, tags, giBaseType, groupFunction) {
    var baseItem;
    if(itemType !== 'Group' && typeof(giBaseType) !== 'undefined') {
        baseItem = itemBuilderFactory.newItemBuilder(giBaseType, itemName + "_baseItem").build()
    }
    if(itemType !== 'Group') {
        groupFunction = undefined;
    }

    var builder = itemBuilderFactory.newItemBuilder(itemType, itemName).
                                                withCategory(category).
                                                withLabel(label).
                                                withGroups(groups);

    builder = builder.withTags(new HashSet(tags));

    if(typeof baseItem !== 'undefined') {
        builder = builder.withBaseItem(baseItem);
    }
    if(typeof groupFunction !== 'undefined') {
        builder = builder.withGroupFunction(groupFunction);
    }

    var item = builder.build();
    managedItemProvider.add(item);
    logDebug("Item added: " + item);
    return item;
}

context.removeItem = function(itemOrItemName) {
    var itemName;

    if(typeof itemOrItemName === 'string') {
        itemName = itemOrItemName;
    } else if(itemOrItemName.hasOwnProperty('name')) {
        itemName = itemOrItemName.name;
    } else {
        logWarn('Item not registered so cannot be removed');
        return null;
    }

    if(getItem(itemName) === null) {
        logWarn('Item not registered so cannot be removed');
        return null;
    }

    managedItemProvider.remove(itemName);

    if(getItem(itemName) === null) {
        logDebug("Item removed: " + itemName);
        return itemName;
    } else {
        logWarn("Failed to remove item: " + itemName);
        return null;
    }
}

})(this);
